yt/yt/server/controller_agent/controllers/multi_job_manager.cpp

```diff
@@ -0,0 +1,194 @@
+        for (int i = 1; i < joblet->CookieGroupSize; i++) {
+            auto guard = TProgressCounterGuard(JobCounter_);
+            guard.SetCategory(EProgressCategory::Pending);
+            replicas->Secondaries.push_back({.ProgressCounterGuard = std::move(guard)});
+        }
+        replicas->MainJobId = joblet->JobId;
+        replicas->Pending = joblet->CookieGroupSize - 1;
+        replicas->NotCompleted = joblet->CookieGroupSize;
+        InsertOrCrash(PendingCookies_, joblet->OutputCookie);
+    } else {
+        auto replicas = GetOrCrash(CookieToReplicas_, joblet->OutputCookie);
+        auto &secondary = replicas->Secondaries[joblet->OutputCookieGroupIndex - 1];
+        secondary.JobId = joblet->JobId;
+        secondary.ProgressCounterGuard.SetCategory(EProgressCategory::Running);
+        joblet->MainJobId = replicas->MainJobId;
+        replicas->Pending--;
+        if (!replicas->Pending) {
+            EraseOrCrash(PendingCookies_, joblet->OutputCookie);
+        }
+    }
+}
+
+bool TMultiJobManager::OnJobCompleted(const TJobletPtr& joblet)
+{
+    if (!IsRelevant(joblet)) {
+        return true;
+    }
+
+    auto replicas = CookieToReplicas_.find(joblet->OutputCookie);
+    if (replicas != CookieToReplicas_.end()) {
+        if (joblet->OutputCookieGroupIndex != 0) {
+            replicas->second->Secondaries[joblet->OutputCookieGroupIndex - 1].ProgressCounterGuard.SetCategory(EProgressCategory::Completed);
+        }
+        --replicas->second->NotCompleted;
+        YT_VERIFY(replicas->second->NotCompleted >= 0);
+        if (!replicas->second->NotCompleted) {
+            CookieToReplicas_.erase(joblet->OutputCookie);
+            return true;
+        }
+    }
+    return false;
+}
+
+std::optional<EAbortReason> TMultiJobManager::ShouldAbortCompletingJob(const TJobletPtr&)
+{
+    return std::nullopt;
+}
+
+bool TMultiJobManager::IsFinished() const
+{
+    return CookieToReplicas_.empty() && PendingCookies_.empty();
+}
+
+TProgressCounterPtr TMultiJobManager::GetProgressCounter() const
+{
+    return JobCounter_;
+}
+
+bool TMultiJobManager::IsRelevant(const TJobletPtr& joblet) const
+{
+    return joblet->CookieGroupSize > 1;
+}
+
+void TMultiJobManager::RegisterMetadata(auto&& registrar)
+{
+    PHOENIX_REGISTER_FIELD(1, CookieToReplicas_);
+    PHOENIX_REGISTER_FIELD(2, PendingCookies_);
+    PHOENIX_REGISTER_FIELD(3, BannedCookies_);
+    PHOENIX_REGISTER_FIELD(4, JobCounter_);
+    PHOENIX_REGISTER_FIELD(5, Task_);
+    PHOENIX_REGISTER_FIELD(6, Logger);
+    PHOENIX_REGISTER_FIELD(7, PendingDataWeight_);
+}
+
+PHOENIX_DEFINE_TYPE(TMultiJobManager);
+
+bool TMultiJobManager::OnUnsuccessfulJobFinish(
+    const TJobletPtr& joblet,
+    EAbortReason abortReason)
+{
+    if (!IsRelevant(joblet)) {
+        // By default after unsuccessful finish of job a cookie is returned to chunk pool.
+        return true;
+    }
+
+    auto replicas = CookieToReplicas_.find(joblet->OutputCookie);
```

Comment:
As I commented above, let use `it` here and local variable `replicas` for the value